File: 1639. Number of Ways to Form a Target String Given a Dictionary/main.py
from typing import List


class Solution:
    def numWays(self, words: List[str], target: str) -> int:
        MOD = 1000000007
        dp = {}
        char_frequency = [[0] * 26 for _ in range(len(words[0]))]

        for i in range(len(words)):
            for j in range(len(words[0])):
                char_frequency[j][ord(words[i][j]) - 97] += 1

        def get_words(words_index, target_index):
            if target_index == len(target):
                return 1
            
            if words_index == len(words[0]) or len(words[0]) - words_index < len(target) - target_index:
                return 0
            
            if (words_index, target_index) in dp:
                return dp[(words_index, target_index)]
            
            count_ways = 0
            cur_pos = ord(target[target_index]) - 97
            
            # skip current word
            count_ways += get_words(words_index + 1, target_index)

            # use current word
            count_ways += char_frequency[words_index][cur_pos] * get_words(words_index + 1, target_index + 1)

            dp[(words_index, target_index)] = count_ways % MOD
            return dp[(words_index, target_index)]
        
        return get_words(0, 0)

        

File: 1639. Number of Ways to Form a Target String Given a Dictionary/main.go
package main

func ord(letter rune) int {
	return int(letter - 'a')
}

func numWays(words []string, target string) int {
	const MOD = 1000000007
	cache := make(map[[2]int]int)
	charFrequency := make([][26]int, len(words[0]), len(words[0]))
	for i := range words {
		for j := range len(words[0]) {
			charFrequency[j][ord(rune(words[i][j]))] += 1
		}
	}

	var getWords func(wordsIndex, targetIndex int) int
	getWords = func(wordsIndex, targetIndex int) int {
		if targetIndex == len(target) {
			return 1
		}
		if wordsIndex == len(words[0]) || len(words[0])-wordsIndex < len(target)-targetIndex {
			return 0
		}

		cacheKey := [2]int{wordsIndex, targetIndex}

		if cachedVal, exists := cache[cacheKey]; exists {
			return cachedVal
		}

		count_ways := 0
		cur_pos := ord(rune(target[targetIndex]))

		// skip current word
		count_ways += getWords(wordsIndex+1, targetIndex)

		// use current word's char
		count_ways += charFrequency[wordsIndex][cur_pos] * getWords(wordsIndex+1, targetIndex+1)

		cache[cacheKey] = count_ways % MOD
		return cache[cacheKey]
	}

	return getWords(0, 0)
}
